### ISSUE TITLE
lang and xml:lang should not use the flat tree

### ISSUE THREAD

**Issue by annevk** [WICG/webcomponents#385]
At least I think they shouldn't. That would mean that a shadow tree can influence the language of children of the shadow host. That seems wrong, no?

**Comment by rniwa** [comment]
Makes sense to me.

**Comment by hayatoito** [comment]
See the past discussion here:
- https://www.w3.org/Bugs/Public/show_bug.cgi?id=27222
- https://github.com/w3c/webcomponents/pull/32/files

@kojiishi
WDYT?

**Comment by annevk** [comment]
`dir` is another attribute I'm a skeptical about for the same reason.

**Comment by kojiishi** [comment]
Yeah, right, for the same reason as `dir`, I think `lang` should inherit.

These two attributes are more environmental, affects text rendering and font fallback even into form fields today in most browsers. Think they're the language setting in your OS which can reach deep into date controls in applications. Or when your app changes the locale of the process, it affects all locales in 3rd party controls in your app.

Also, technically, `dir` is mapped to the CSS `direction` property, and CSS inherits composed tree, so `dir` and `direction` behaving differently is problematic. Blink and WebKit maps `lang` to CSS `-webkit-locale` property so the same technical constrain applies, though it's not a standard property.

**Comment by annevk** [comment]
But how do you know whether the component uses the same language and direction? What if that changes? I think this again largely depends on the use case for the component.

**Comment by kojiishi** [comment]
The same way as how apps/controls work today? They can override if they want to, but defaults to OS/process locale. You want a calendar in an app to honor your OS setting by default, don't you?

Note, the CSS `all: initial` does not reset `direction` because `all` excludes `direction` and `unicode-bidi` specifically ([spec](https://drafts.csswg.org/css-cascade-4/#propdef-all)). I see similarity here.

**Comment by annevk** [comment]
@kojiishi if it's just a control that might be okay (though in that case the control will also need to be able to adapt based on the language and direction inputs it gets), but if it's also content it gets a little weird. Now perhaps if you include content the advice would be to always include these attributes yourself since you might get overwritten, but if the majority is en/ltr that will not get noticed much.

**Comment by kojiishi** [comment]
Can you elaborate how weird it is for you? It looks to me that contents is also weird if it does not inherit. Let's say a page declares `lang="ja"` without specifying `font-family` and my browser picked MSGothic to render all characters including Latin alphabets. Shouldn't English content in the page pick the same consistent font, unless it overrides font or lang? By "environmental" I meant these two attributes can change browsers' default behavior. Components relying on browsers' default need to inherit these two attributes to look consistent. Of course components can pick its own font or lang or dir, in which case they should be honored.

When en/ltr content is inserted as a component into ar/rtl document without declaring lang/dir, they should have ar/rtl. Font fallback and bidi reordering algorithm can take care of such situation. There are some situation where the algorithm cannot determine the correct answer and produce weird results, but that's rather edge cases compared to weird results by not inheriting.

If an application does not honor RTL setting in the OS, it's not usable. If an application honors RTL settings but produces some weird behavior when using on RTL OS, it's a bug. CJK users know poorly localized apps render (c) mark as Katakana on CJK OS, it's a bug, but if it renders all CJK characters as Tofu, it's not usable.

I mean, I agree that there are some cases where not inheriting is better, but I think not inheriting will result in much larger issues. If you have cases that inheriting can cause critical issues such as not-usable, it'd be great to know.

Also, although it's less important than seeking for the right answer, the technical constrain I mentioned earlier is not really ignorable.

**Comment by annevk** [comment]
I think there's three kind of scenarios:
1. Component matches the language of the page so this doesn't matter. Component developers are unlikely to set lang/dir. This is probably the majority case.
2. Component does not match the language of the page. Given 1, this will be problematic since now screen readers etc. think the component is in the language of the page while it isn't.
3. Component uses the language of the page to render different strings to the user. For this case the default doesn't really matter and the component is best of querying the host element.

Basically, I think by catering to 1 by inheriting by default, we hurt 2. But maybe you're right and this is okay. Hmm.

**Comment by kojiishi** [comment]
> by catering to 1 by inheriting by default, we hurt 2

2 can either be good or problematic, but I think "good" cases win over "problematic" cases.
1. Inheriting is problematic when showing English text in an editor component in a French page because spellcheck thinks it's French.
2. Not inheriting is problematic when showing non-English text in an editor component in any pages because spellcheck thinks it's English when lang is not set.

1 is predictable and I can afford to, that happens when I pasted different languages into my plain text editor too.

2 looks hopeless. All such language dependent components would not be usable outside of English unless component authors explicitly support getting lang from the shadow host.

Spellcheck for one example, but there are others, such as underline or [text emphasis](https://drafts.csswg.org/css-text-decor-3/#text-emphasis-position-property). I can accept Japanese text is underlined like Chinese if the page is Chinese, but all text in shadow trees in any pages being Chinese isn't really acceptable.

**Comment by rniwa** [comment]
A good discussion.  I agree with @kojiishi here that inheriting the locale is probably a good default behavior.

**Comment by annevk** [comment]
Okay. (I'd like the lang -> `-webkit-locale` thing to be standardized at some point.)
